lib/optimizely/project_config.rb

```diff
@@ -246,19 +253,119 @@ def get_variation_id_from_key(experiment_key, variation_key)
+      # check for nil and empty string experiment ID
+      if experiment_id.nil? or experiment_id.empty?
+        # this case is logged in get_experiment_from_key
+        return nil
+      end
+
+      unless experimentToVariationMap[experiment_id].present?
+        @logger.log(Logger::DEBUG, "No experiment '#{experiment_key}' mapped to user '#{user_id}' in the forced variation map.")
+        return nil
+      end
+
+      variation_id = experimentToVariationMap[experiment_id]
+      # check for nil and empty string variation ID
+      if variation_id.nil? or variation_id.empty?
+        @logger.log(Logger::DEBUG, "No variation mapped to experiment '#{experiment_key}' in the forced variation map.")
+        return nil
+      end
+
+      variation_key = ""
+      variation_id_map = @variation_id_map[experiment_key]
+      if variation_id_map
+        variation = get_variation_from_id(experiment_key,variation_id)
+        variation_key = variation["key"] if variation
+      end
+
+      # check if the variation exists in the datafile
+      if variation_key.empty?
+        # this case is logged in get_variation_from_id
+        return nil
+      end
+
+      @logger.log(Logger::DEBUG, "Variation '#{variation_key}' is mapped to experiment '#{experiment_key}' and user '#{user_id}' in the forced variation map")
+      
+      variation
+    end
+
+    def set_forced_variation(experiment_key, user_id, variation_key)
+      # Sets a Hash of user IDs to a Hash of experiments to forced variations.
+      #
+      # experiment_key - String Key for experiment.
+      # user_id - String ID for user.
+      # variation_key - String Key for variation. If null, then clear the existing experiment-to-variation mapping.
+      #
+      # Returns a boolean value that indicates if the set completed successfully.
+
+      #  check for null and empty string user ID
+      if user_id.nil? or user_id.empty?
+        @logger.log(Logger::DEBUG, "User ID is invalid")
+        return FALSE
+      end
+
+      experiment = get_experiment_from_key(experiment_key)
+      experiment_id = experiment["id"] if experiment
+      #  check if the experiment exists in the datafile
+      if experiment_id.nil? or experiment_id.empty?
+        return FALSE
+      end
+
+      #  clear the forced variation if the variation key is null
+      if variation_key.nil? or variation_key.empty?
+        @forced_variation_map[user_id].delete(experiment_id) if @forced_variation_map[user_id].present?
+        @logger.log(Logger::DEBUG, "Variation mapped to experiment '#{experiment_key}' has been removed for user '#{user_id}'.")
+        return TRUE
+      end
+
+      variation_id = get_variation_id_from_key(experiment_key, variation_key)
+
+      #  check if the variation exists in the datafile
+      unless variation_id
+        #  this case is logged in get_variation_id_from_key
+        return FALSE
+      end
+
+      @forced_variation_map[user_id][experiment_id] = variation_id
```

Comment:
can u plz. check if forced_variation_map doesn't have any dictionary for usreid key, will it create automatically or you will need to explicitly create key, dictionary first and then [experimentid, variationid]?